Add spatial types

ddegasperi opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | #7081 

#### Summary

This PR introduces spatial data type support for DBAL, starting with PostgreSQL/PostGIS and MySQL. This implementation is intended to serve as a reference and guide for adding support for other database platforms in the future.

#### Core Spatial Types

- **GeometryType** (`Types::GEOMETRY`) - For planar coordinate systems
- **GeographyType** (`Types::GEOGRAPHY`) - For spherical earth coordinates
- Both types handle GeoJSON format for data exchange

#### Schema API Integration

- Extended `Column` and `ColumnEditor` with `geometryType` and `srid` properties
- Clean fluent API: `Column::editor()->setGeometryType('POINT')->setSrid(4326)`
- Full PostgreSQL platform support with PostGIS-specific SQL generation
- Schema introspection and table creation/modification support

#### What's next

- [x] Discuss the PR
- [X] Add other platforms (PostgreSQL/PostGIS, MySQL)
- [X] Add spatial index (PostgreSQL/PostGIS, MySQL already support creation of spatial indexes)

ddegasperi committed:
Core spatial types implementation

Add geometry and geography types for spatial data

Introduces new spatial data types for handling geometric and geographic data in database applications. GeometryType handles planar coordinates while GeographyType handles spherical earth coordinates, both using GeoJSON format.

This provides foundation for spatial data operations across database platforms that support spatial extensions.

ddegasperi committed:
Core spatial types implementation

Add geometry and geography types for spatial data

Introduces new spatial data types for handling geometric and geographic data in database applications. GeometryType handles planar coordinates while GeographyType handles spherical earth coordinates, both using GeoJSON format.

This provides foundation for spatial data operations across database platforms that support spatial extensions.

ddegasperi committed:
Add PostgreSQL spatial type support

Extends Column and ColumnEditor with geometryType and srid properties to support PostgreSQL's PostGIS spatial types and provides a clean schema API for working with GEOMETRY and GEOGRAPHY columns while maintaining backward compatibility.

This builds on the core spatial types implementation to complete the PostgreSQL spatial type support at the schema level.

ddegasperi committed:
Add PostgreSQL spatial type support

Extends Column and ColumnEditor with geometryType and srid properties to support PostgreSQL's PostGIS spatial types and provides a clean schema API for working with GEOMETRY and GEOGRAPHY columns while maintaining backward compatibility.

This builds on the core spatial types implementation to complete the PostgreSQL spatial type support at the schema level.

derrabus commented on tests/Platforms/PostgreSQLPlatformTest.php:
```diff
@@ -1060,4 +1060,208 @@ public function testAlterTableChangeJsonbToJson(): void
+            )
+            ->create();
+
+        self::assertSame(
+            [
+                'CREATE TABLE spatial_table (id INT NOT NULL, '
+                . 'location geometry(geometry) NOT NULL, '
+                . 'point_location geometry(point,4326) NOT NULL, '
+                . 'polygon_area geometry(polygon) NOT NULL)',
+            ],
+            $this->platform->getCreateTableSQL($table),
+        );
```
Please focus on functional tests. The kind of test that you're building here is pretty much worthless because it does not tell us if the SQL actually works when it hits a database.

ddegasperi replied:
Thanks for the feedback! I already had in mind to add functional tests, since they’re definitely the right way to validate the SQL.
Because these tests require PostgreSQL with the PostGIS extension, would you suggest:

- creating a dedicated GitHub workflow (e.g. `postgresql-postgis.yml`) that uses the `postgis/postgis` image, or
- replacing the current PostgreSQL workflow with that image so that all PostgreSQL jobs run with PostGIS enabled?

I’d be happy to go with whichever approach best fits the project’s CI strategy.

derrabus replied:
We need to make sure DBAL still works with Postgres databases that don't have these extensions installed. So it's probably best for now if we add an additional check for Postgres that uses the latest postgis image and leave the other checks as they are.

ddegasperi replied:
@derrabus I’ve added the functional tests as suggested and pushed them. However, I noticed that none of the functional tests actually executed in the PR pipeline — is there something I need to do to make sure they run?

If you already had a chance to take a look, I’d really appreciate some feedback on whether this is heading in the right direction. That way I can adjust before proceeding with support for a second platform (e.g. MySQL).

derrabus replied:
As far as I can tell, your new workflows were run: https://github.com/doctrine/dbal/actions/runs/17288021054/job/49069200071?pr=7096

ddegasperi replied:
Yes, I noticed an error in `continuous-integration.yml` only afterwards. After fixing it, the functional tests were executed correctly.

I have one question regarding the functional tests: in MySQL there is a test case `Doctrine\DBAL\Tests\Functional\Schema\MySQLSchemaManagerTest::testColumnIntrospection` that attempts to define any type as a column to compare it against the same definition. At the moment, I’ve introduced two types for PostgreSQL/PostGIS, `geometry` and `geography`. Since `geography` is not available in MySQL, the current test setup would fail.

Is this behavior intentional, or would it be acceptable to adjust the tests so that they explicitly check which types are supported on MySQL?

ddegasperi committed:
Add functional test cases for spatial type support in PostgreSQL

This commit adds functional testing for PostGIS spatial types (GEOMETRY and GEOGRAPHY) with schema introspection and CI integration.

The implementation leverages PostgreSQL's native type system for introspection, making it compatible with any PostgreSQL instance without requiring PostGIS system tables to be accessible during schema operations.

ddegasperi committed:
Add functional test cases for spatial type support in PostgreSQL

This commit adds functional testing for PostGIS spatial types (GEOMETRY and GEOGRAPHY) with schema introspection and CI integration.

The implementation leverages PostgreSQL's native type system for introspection, making it compatible with any PostgreSQL instance without requiring PostGIS system tables to be accessible during schema operations.

derrabus commented on src/Types/GeometryType.php:
```diff
@@ -0,0 +1,69 @@
+            return null;
+        }
+
+        if (is_string($value)) {
+            return $value;
+        }
```
Is there any object representation that we could hydrate this value to? If we just pass on the strings we might end up with a leaky abstraction.

ddegasperi replied:
Good point — I agree that passing raw strings would be a leaky abstraction. Since the implementation uses WKT, I could introduce a lightweight `WKT` value object that wraps and optionally validates the string (e.g. via regexp). This would give us a clear object representation without pulling in a full geometry library. Would this be in line with what you had in mind?

ddegasperi replied:
I decided not to introduce a WKT value object and went with a GeoJSON value object instead. GeoJSON provides both the geometry and its metadata (such as SRID) in a consistent way across different database platforms.

tibobaldwin commented:
Adding spatial index management would be interesting for this PR I think :).

For instance:
`CREATE INDEX mytable_geom_x ON mytable USING GIST (geom)`

Sources:
- https://postgis.net/documentation/faq/spatial-indexes/
- https://postgis.net/workshops/postgis-intro/indexing.html
- https://postgis.net/docs/using_postgis_dbmanagement.html#build-indexes

ddegasperi commented:
> Adding spatial index management would be interesting

Thanks for the advice — I agree, index management is definitely an important aspect when working with geometry data. I haven’t investigated it in depth yet, but it looks like this might already be supported by specifying the index type. A quick search in the repository shows an `IndexType` enum with a `SPATIAL` option, so I’ll take a closer look at how that could be integrated here.

ddegasperi committed:
Add MySQL spatial type support

Extends AbstractMySQLPlatform with GEOMETRY type support and enhances MySQLSchemaManager to introspect spatial columns with geometryType and SRID properties.

MySQL supports GEOMETRY types (POINT, LINESTRING, POLYGON, etc.) with optional SRID constraints using the conditional comment syntax for MySQL 8.0.3+.

ddegasperi committed:
Add MySQL spatial type support

Extends AbstractMySQLPlatform with GEOMETRY type support and enhances MySQLSchemaManager to introspect spatial columns with geometryType and SRID properties.

MySQL supports GEOMETRY types (POINT, LINESTRING, POLYGON, etc.) with optional SRID constraints using the conditional comment syntax for MySQL 8.0.3+.

ddegasperi committed:
Introduce Geometry value objects to eliminate leaky abstractions

Refactor GeometryType and GeographyType to replace direct GeoJSON string handling with dedicated value objects, preventing exposure of database-specific formats to application code.

This commit introduces a Geometry value object that encapsulates a supporting GeoJSON value object responsible for validating and wrapping GeoJSON representations.

ddegasperi committed:
Introduce Geometry value objects to eliminate leaky abstractions

Refactor GeometryType and GeographyType to replace direct GeoJSON string handling with dedicated value objects, preventing exposure of database-specific formats to application code.

This commit introduces a Geometry value object that encapsulates a supporting GeoJSON value object responsible for validating and wrapping GeoJSON representations.

ddegasperi committed:
Add spatial index support for PostgreSQL/PostGIS

Implement spatial index creation and introspection for PostgreSQL using the GIST (Generalized Search Tree) index method — the standard access method for spatial data in PostGIS.
This commit introduces a new getIndexMethodSQL() hook in AbstractPlatform for platform-specific index clauses, and overrides it in PostgreSQLPlatform to emit "USING GIST" for spatial indexes.

SQL generation examples:
MySQL → CREATE SPATIAL INDEX idx ON table (col)
PostgreSQL → CREATE INDEX idx ON table USING GIST (col)

ddegasperi committed:
Add spatial index support for PostgreSQL/PostGIS

Implement spatial index creation and introspection for PostgreSQL using the GIST (Generalized Search Tree) index method — the standard access method for spatial data in PostGIS.
This commit introduces a new getIndexMethodSQL() hook in AbstractPlatform for platform-specific index clauses, and overrides it in PostgreSQLPlatform to emit "USING GIST" for spatial indexes.

SQL generation examples:
MySQL → CREATE SPATIAL INDEX idx ON table (col)
PostgreSQL → CREATE INDEX idx ON table USING GIST (col)

jungleman12 commented:
Wow, that's exactly what I'd like to see integrated into Doctrine soon. Great work!

ddegasperi commented:
Hi @derrabus 

I’ve addressed all the feedback received so far and updated the PR accordingly.
When you get a chance, could you please let me know if there’s anything else I can do to help move this forward?

Thanks a lot for your time

derrabus commented on .github/workflows/continuous-integration.yml:
```diff
@@ -135,6 +135,31 @@ jobs:
+      matrix:
+        php-version:
+          - "8.3"
+          - "8.4"
```
```suggestion
          - "8.5"
```

I don't expect any new insights from testing on multiple PHP versions, do you? Let's not bloat the matrix too much. Let's test with the lastest PHP version on both extensions and on the lowest supported PHP version with one extension only.

ddegasperi replied:
Applied as suggested.

derrabus commented on .github/workflows/phpunit-postgis.yml:
```diff
@@ -0,0 +1,64 @@
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
```
```suggestion
        uses: actions/checkout@v6
```

ddegasperi replied:
Applied as suggested.

derrabus commented on src/Platforms/AbstractMySQLPlatform.php:
```diff
@@ -219,6 +221,65 @@ public function getBooleanTypeDeclarationSQL(array $column): string
+
+        if ($srid !== null) {
+            // MySQL 8.0.3+ supports SRID attribute using conditional comment syntax
+            $sql .= sprintf(' /*!80003 SRID %d */', $srid);
```
Do we really need to emit those version-conditional comments? We should know which version of MySQL we're connecting to.

ddegasperi replied:
Extracted the logic into the platform classes.
Since MySQL 5.7 doesn’t support SRIDs, the actual SRID handling is implemented only in MySQL80Platform without conditional comment.

derrabus commented on src/Schema/MySQLSchemaManager.php:
```diff
@@ -355,6 +395,12 @@ protected function selectTableColumns(string $databaseName, ?string $tableName =
+        // For MariaDB, SRID is parsed from COLUMN_TYPE conditional comments
+        $sridColumn = $this->platform instanceof MariaDBPlatform
+            ? 'NULL AS srs_id'
+            : 'c.SRS_ID AS srs_id';
```
Please don't perform `instanceof` checks on the platform. Move this logic into the platform classes instead.

ddegasperi replied:
Extracted the logic into the platform classes.
Since MySQL 5.7 doesn’t support SRIDs, the actual SRID handling is implemented only in MySQL80Platform.

derrabus commented on src/Types/GeoJSON.php:
```diff
@@ -0,0 +1,224 @@
+        try {
+            return json_encode($this->data, JSON_THROW_ON_ERROR);
+        } catch (JsonException $e) {
+            throw new InvalidArgumentException(
```
How can a method without arguments raise errors on an "invalid argument"?

ddegasperi replied:
Replaced the invalid argument exception with a `RuntimeException`.

derrabus commented on src/Connection.php:
```diff
@@ -538,6 +544,67 @@ private function extractTypeValues(array $columns, array $types): array
         return $typeValues;
     }
 
+    /**
+     * Normalizes types array from positional or associative to associative format.
+     *
+     * @param array<int<0,max>, string|ParameterType|Type>|array<string, string|ParameterType|Type> $types
+     * @param list<string>                                                                           $columnNames
+     *
+     * @return array<string, string|ParameterType|Type>
+     */
+    private function normalizeTypes(array $types, array $columnNames): array
+    {
```
Why do we have to do this?

ddegasperi replied:
I introduced `normalizeTypes()` because I needed a reliable way to support SQL-level type conversion via `Type::convertToDatabaseValueSQL()`, which is required for spatial types and may also be relevant for other custom types in the future.

With `insert()` and `update()`, the `$types` argument can be passed in two different forms:
- Positional, e.g. `[0 => 'geometry', 1 => 'string']`
- Associative, e.g. `['location' => 'geometry', 'name' => 'string']`

The new `getPlaceholderForColumn()` method operates at the column level and needs to know whether a specific column requires SQL-level conversion using `convertToDatabaseValueSQL()`. For that decision, the type must be known by column name.

When `$types` is positional, there is no straightforward or safe way to determine the type for a given column name at that point. I added `normalizeTypes()` to normalize both positional and associative `$types` into a single, column-name–keyed structure. This allows `getPlaceholderForColumn()` to consistently determine:
- which type applies to each column
- whether SQL-level conversion should be applied

derrabus replied:
So the current type abstraction does not work for these new types? It's quite unusual that we need to change the connection class on the wrapper layer for introducing new types.

ddegasperi replied:
Thanks for the discussion — I did a bit more digging to better understand how type conversion is currently applied.

In DBAL itself, `convertToDatabaseValue()` is called when binding parameters, so value-level conversion works as expected. However, I couldn’t find any code paths in DBAL where `convertToDatabaseValueSQL()` (or `convertToPHPValueSQL()`) is actually invoked. As far as I can tell, these methods are currently required and used by `doctrine/orm`, not by DBAL directly.

This becomes relevant for spatial types because converting GeoJSON into a native geometry value in PostGIS, MySQL, or MariaDB cannot be done via casting — all of them require an explicit SQL function call such as `ST_GeomFromGeoJSON(...)`. If the SQL-level conversion hook isn’t invoked, any custom SQL logic implemented in `convertToDatabaseValueSQL()` would effectively be ignored.

From this perspective, DBAL currently:
* supports value-level conversion at bind time (`convertToDatabaseValue()`),
* but does not apply SQL-level conversion hooks during `insert()` / `update()` flows.

That’s the context in which I explored a possible solution at the connection layer: without a place where `convertToDatabaseValueSQL()` is actually used, this appeared to be a way to express database-specific SQL transformations while still allowing users to work with a user-friendly input format like GeoJSON.

Happy to adjust the approach if there’s a more idiomatic way to support this in DBAL — I mainly wanted to share these findings to provide some additional context around the motivation.

derrabus replied:
So there's no binary format that can be insterted directly into those columns? I've run into similar problems while working on the `VECTOR` type. And you're right, `convertToDatabaseValueSQL()` is never called by the DBAL, currently. Maybe we need a better abstraction here.

ddegasperi replied:
Good question — this does indeed seem very similar to the VECTOR case.

While geometry columns can sometimes accept WKB directly, conversion functions like `ST_GeomFromWKB()` or `ST_GeomFromGeoJSON()` are generally the more robust and recommended approach across spatial databases, as they make the conversion explicit and behave consistently.

There are also some portability concerns with a pure WKB-based approach:
* **EWKB (with embedded SRID) is a PostGIS extension** and isn’t supported by MySQL or MariaDB, so SRID handling would still require additional logic.
* Relying on implicit casting of binary values feels more fragile than using explicit SQL functions, especially across different platforms and versions.

Because of that, a WKB-based workaround could work in some cases, but it feels more like a workaround than a solid abstraction.

Given this, I’d appreciate some guidance on direction:
* should we explore a WKB-based value-level approach despite these trade-offs, or
* is it worth discussing a DBAL-level abstraction for types that require SQL-level transformation during insert/update (which would also apply to cases like VECTOR)?

Personally, I would lean towards exploring a proper abstraction if that aligns with DBAL’s design goals, but I’m happy to follow the direction you think makes the most sense.

derrabus replied:
> is it worth discussing a DBAL-level abstraction for types that require SQL-level transformation during insert/update

I think so, yes.

ddegasperi replied:
Before going deeper into an implementation, I’d like to better understand what kind of abstraction you have in mind. From my side, I was trying to reason about where SQL-level type conversion could live in a way that stays consistent and avoids duplication.

One idea I explored was whether it would make sense for `Connection::insert()` / `update()` to internally rely on `QueryBuilder`, so that SQL-level conversion (via `convertToDatabaseValueSQL()`) could be handled in a single place rather than re-implemented in multiple code paths.

That said, I’m not attached to this approach — I’d be very interested to hear your thoughts on:
- what shape you imagine this abstraction taking, and
- whether using `QueryBuilder` inside `Connection` would align with DBAL’s architectural direction, or if you’d prefer to keep them clearly independent.

derrabus commented on tests/Functional/Schema/MySQL/SpatialTypesTest.php:
```diff
@@ -0,0 +1,378 @@
+                    ->setUnquotedName('geom')
+                    ->setTypeName(Types::GEOMETRY)
+                    ->setGeometryType('POINT')
+                    ->setSrid(4326)
```
Is this magic number documented anywhere?

ddegasperi replied:
Extracted and documented the magic numbers (SRID codes) in a shared trait `SpatialReferenceSystems` used by the spatial tests.

derrabus commented on tests/Functional/Schema/MySQL/SpatialTypesTest.php:
```diff
@@ -0,0 +1,378 @@
+        // MariaDB does not support SRID specification in column definitions
+        if (! $this->connection->getDatabasePlatform() instanceof MariaDBPlatform) {
+            self::assertSame(4326, $sridGeom->getSrid());
+        }
```
So what is the expected behavior for MariaDB then?

ddegasperi replied:
Split the tests between MariaDB and MySQL. Since MariaDB doesn’t support SRIDs, the expected behavior there is asserting a `null` SRID.

derrabus commented on tests/Functional/Schema/PostgreSQL/PostGISTest.php:
```diff
@@ -0,0 +1,400 @@
+        // Execute ALTER TABLE
+        foreach ($alterSql as $sql) {
+            $this->connection->executeStatement($sql);
+        }
```
The schema manager can apply the diff for us. We don't have to feed the SQL statements into the connection ourselves.

ddegasperi replied:
Applied as suggested.

derrabus commented on tests/TestUtil.php:
```diff
@@ -147,6 +148,23 @@ private static function initializeDatabase(): void
+                }
+
+                $testConn->close();
+            }
```
Isn't this something our CI setup should do instead?

ddegasperi replied:
Moved this logic into the CI setup as suggested.

derrabus reviewed (changes requested):
Thank your for all the work you've put into this PR. Your changes look very promising.

ddegasperi commented:
> Thank your for all the work you've put into this PR. Your changes look very promising.

Thanks a lot for your review and the detailed feedback!
I really appreciate the time you took to go through the PR.
I’ll work on the requested changes as soon as possible and hopefully be able to address everything within the next week.

derrabus commented on tests/SpatialReferenceSystems.php:
```diff
@@ -0,0 +1,45 @@
+ *
+ * These EPSG codes define coordinate reference systems used across spatial functionality tests.
+ */
+trait SpatialReferenceSystems
```
I've never seen a trait that only holds constants tbh. Let's make this a class instead?

ddegasperi replied:
Converted it into a class as suggested.

derrabus commented:
The tests fail on MySQL.

ddegasperi commented:
> The tests fail on MySQL.

Apologies for that — I hadn’t rerun the tests on MySQL 5.7 yet. I’ve now applied the necessary fixes so the test suite passes on both MySQL 5.7 and MySQL ≥ 8, and I’ve verified that the static analysis checks pass as well.

derrabus commented:
> the test suite passes on both MySQL 5.7 and MySQL ≥ 8

No, it still fails on 5.7.

ddegasperi commented:
Just a gentle follow-up on this topic. I completely understand things get busy — I just wanted to check whether you’ve had a chance to think about the abstraction question around SQL-level type conversion.

I’m happy to move forward in whichever direction you think fits best (e.g. exploring a QueryBuilder-based approach, or something different). If you have a preference, that would help me focus the implementation accordingly.

seb-jean commented:
Hi @ddegasperi, there is still an error related to PHPStan in the pipeline, but I am unsure if it is related to your PR.

ddegasperi commented:
> Hi @ddegasperi, there is still an error related to PHPStan in the pipeline, but I am unsure if it is related to your PR.

@seb-jean The reported PHPStan error points to `src/Schema/ColumnDiff.php`, which is not touched by this PR. The issue appears to be triggered by the introduction of the PHP 8.5 checks and was already addressed in the base branch by @derrabus (via an update to the PHPStan baseline).

I’ve rebased my branch on the latest `4.5.x` branch including that fix, so the pipeline should now pass on the next run.

seb-jean commented:
That's exactly what I thought when I went through PR's code.
Thank you for your quick response and explanation.

seb-jean commented on src/Platforms/PostgreSQLPlatform.php:
```diff
@@ -825,11 +827,70 @@ public function getJsonbTypeDeclarationSQL(array $column): string
+    public function getGeographyTypeDeclarationSQL(array $column): string
+    {
+        $geometryType = $column['geometryType'] ?? 'geometry';
+        $srid         = $column['srid'] ?? null;
```
```suggestion
        $srid         = $column['srid'] ?? 4326;
```

ddegasperi replied:
Applied as suggested

seb-jean commented on src/Types/Geometry.php:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @throws InvalidArgumentException If the GeoJSON format is invalid.
+     */
+    public static function fromGeoJSON(string $json): self
```
I think I need to do something like this to add a Point type data point:
```php
$location->setCoordinates(Geometry::fromGeoJSON((string) json_encode([
   'type' => 'Point',
   'coordinates' => [$location->getLongitude(), $location->getLatitude()],
   'crs' => ['type' => 'name', 'properties' => ['name' => 'EPSG:4326']],
])));
```

I'm finding the average DX, but I don't think there's another way to do it?

ddegasperi replied:
Good point — I agree the current DX isn't ideal. Having to hand-build the array, `json_encode` it, and spell out the full `crs` block just to store a single Point is more ceremony than it should be.

As a first, lightweight option, I was thinking of adding named constructors on the `Geometry` value object covering the RFC 7946 types (a sketch is at the bottom of this comment). They would all funnel through the existing `fromGeoJSON`/`toGeoJSON` internals, so the wire format stays GeoJSON and the round-trip keeps working through a single bind parameter (`ST_GeomFromGeoJSON(?)`, with the SRID carried inside the payload). The factories would purely be an ergonomic front door — nothing changes underneath.

I deliberately leaned away from a full typed hierarchy (`Point`, `LineString`, … as separate classes). It would be the richer model, but a complete geometry type system — coordinate validation, Z/M dimensions, accessors, equality — feels more like the domain of dedicated third-party geometry libraries, and I'd be cautious about DBAL reinventing and then having to maintain that surface long-term.

Before going further, it would really help to hear from your side:

- What does your data usually look like — raw coordinates like the sketch below, or do you already have WKT/EWKT strings on hand? The former fits these factories nicely; the latter would point more towards a PHP-side `fromWKT()` that parses into the same GeoJSON-backed object rather than touching the wire format.
- Would constructors like these actually smooth out your workflow, or is there something else in the day-to-day usage that feels awkward?

This is just an early idea, and the final shape of the public API would of course be up to the maintainers — but I'd like to make sure whatever we end up proposing genuinely solves the use case you ran into.

---

For reference, here's how the constructors might look:

```php
// Point: longitude, latitude
Geometry::point(11.34, 46.49, srid: 4326);

// LineString: a list of [lon, lat] positions
Geometry::lineString([
    [11.34, 46.49],
    [11.36, 46.50],
    [11.38, 46.48],
], srid: 4326);

// Polygon: a list of linear rings — the first is the exterior ring,
// any further rings are holes. Each ring is closed (first position === last).
Geometry::polygon([
    [[11.0, 46.0], [11.5, 46.0], [11.5, 46.5], [11.0, 46.5], [11.0, 46.0]], // exterior
    [[11.1, 46.1], [11.2, 46.1], [11.2, 46.2], [11.1, 46.1]],               // hole (optional, repeatable)
], srid: 4326);

// MultiPoint: a list of points
Geometry::multiPoint([
    [11.34, 46.49],
    [11.40, 46.51],
], srid: 4326);

// MultiLineString: a list of LineStrings
Geometry::multiLineString([
    [[11.34, 46.49], [11.36, 46.50]],
    [[11.40, 46.51], [11.42, 46.52]],
], srid: 4326);

// MultiPolygon: a list of Polygons (each Polygon being a list of rings:
// exterior ring first, optional holes after)
Geometry::multiPolygon([
    [
        [[11.0, 46.0], [11.5, 46.0], [11.5, 46.5], [11.0, 46.5], [11.0, 46.0]], // polygon 1 — exterior
        [[11.1, 46.1], [11.2, 46.1], [11.2, 46.2], [11.1, 46.1]],               // polygon 1 — hole
    ],
    [
        [[12.0, 47.0], [12.5, 47.0], [12.5, 47.5], [12.0, 47.0]],               // polygon 2 — exterior only
    ],
], srid: 4326);

// GeometryCollection: a list of Geometry objects
Geometry::collection([
    Geometry::point(11.34, 46.49),
    Geometry::lineString([[11.34, 46.49], [11.36, 46.50]]),
], srid: 4326);
```

seb-jean replied:
For me, it's simple: I have the latitude and longitude of different cities, for example, and I want to insert them into a database.

seb-jean commented on .github/workflows/continuous-integration.yml:
```diff
@@ -130,26 +130,51 @@ jobs:
+        extension:
+          - "pgsql"
+          - "pdo_pgsql"
+        include: # understand what this include does!
```
```suggestion
        include:
```

ddegasperi replied:
Applied as suggested

seb-jean commented on src/Platforms/MariaDBPlatform.php:
```diff
@@ -58,6 +58,12 @@ public function getColumnTypeSQLSnippet(string $tableAlias, string $databaseName
+    public function getSridColumnSQL(): string
+    {
+        // MariaDB does not support SRS_ID column in information_schema.COLUMNS
+        return 'NULL AS srs_id';
```
The `getSridColumnSQL()` override in `MariaDBPlatform` is redundant - `AbstractMySQLPlatform` already returns `'NULL AS srs_id'`, so this method can be removed.

ddegasperi replied:
Applied as suggested

seb-jean commented on src/Platforms/PostgreSQLPlatform.php:
```diff
@@ -825,11 +827,70 @@ public function getJsonbTypeDeclarationSQL(array $column): string
+     */
+    public function getGeographyTypeDeclarationSQL(array $column): string
+    {
+        $geometryType = $column['geometryType'] ?? 'geometry';
```
```suggestion
        $geometryType = $column['geometryType'] ?? 'geography';
```

ddegasperi replied:
Thanks for flagging it. I looked into this, and I believe the current `'geometry'` default is actually correct, even though I agree it reads like a copy-paste slip at first glance.

The `geometryType` here is the PostGIS *subtype* that goes inside the modifier - `geography(<subtype>, <srid>)`. The valid values are the shared geometry subtypes: `GEOMETRY` (the generic "any type" placeholder), `POINT`, `LINESTRING`, `POLYGON`, etc. There's no subtype called `geography` — that's the column type, not a subtype. So:

- `geography(geometry, 4326)` - generic geography column (what we emit by default)
- `geography(geography, 4326)` - PostGIS rejects this

Defaulting to `'geometry'` gives us the generic "any geography type" column, which is what we want when no subtype is specified.

seb-jean replied:
Okay, that's very clear, thank you!

seb-jean commented on src/Platforms/PostgreSQLPlatform.php:
```diff
@@ -825,11 +827,70 @@ public function getJsonbTypeDeclarationSQL(array $column): string
         return 'JSONB';
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    public function getGeometryTypeDeclarationSQL(array $column): string
+    {
+        $geometryType = $column['geometryType'] ?? 'geometry';
+        $srid         = $column['srid'] ?? null;
+
+        return sprintf('geometry(%s%s)', strtolower($geometryType), $srid !== null ? ',' . $srid : '');
+    }
```
```suggestion
    public function getGeometryTypeDeclarationSQL(array $column) : string {
        $geometryType = $column['geometryType'] ? ? null;
        $srid = $column['srid'] ? ? null;
    
        if ($geometryType == = null&& $srid == = null) {
            return 'geometry';
        }
    
         return sprintf(
             'geometry(%s%s)',
             strtolower($geometryType ?? 'geometry'),
             $srid !== null ? ',' . $srid : '',
         );
    }
```

ddegasperi replied:
Applied as suggested

seb-jean commented:
> @seb-jean I've finally had time to come back to this PR and work through the open points.
> 
> I've now applied most of the suggestions from your review, and I believe the remaining feedback is addressed as well. There are still a couple of points around developer experience (e.g. how geometry values are constructed) that I've left open for discussion, since I'd like to settle on the right direction together before implementing them.
> 
> Whenever you get a chance, I'd really appreciate it if you could take another look.

I find the DX regarding the use of Geometry:: interesting. Thank you :)

Another thing, I haven't seen the documentation.

ddegasperi committed:
Document spatial type support

Spatial types need more than a standard type entry because they
introduce the Geometry/GeoJSON value objects. The bulk goes in a new
reference/spatial-types.rst guide, with concise geometry and geography
entries in reference/types.rst cross-linking to it.
